service/src/main/java/bio/terra/drshub/tracking/TrackingInterceptor.java

```diff
@@ -0,0 +1,103 @@
+package bio.terra.drshub.tracking;
+
+import bio.terra.bard.model.EventProperties;
+import bio.terra.common.iam.BearerTokenFactory;
+import bio.terra.drshub.services.TrackingService;
+import bio.terra.drshub.util.AsyncUtils;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import lombok.extern.slf4j.Slf4j;
+import org.jetbrains.annotations.NotNull;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Component;
+import org.springframework.web.method.HandlerMethod;
+import org.springframework.web.servlet.HandlerInterceptor;
+import org.springframework.web.util.ContentCachingRequestWrapper;
+
+/**
+ * Class to intercept requests and log them to the tracking service (Bard). Methods annotated with
+ * the {@link TrackCall} annotation will be tracked.
+ */
+@Component
+@Slf4j
+public record TrackingInterceptor(
+    TrackingService trackingService,
+    AsyncUtils asyncUtils,
+    BearerTokenFactory bearerTokenFactory,
+    ObjectMapper objectMapper)
+    implements HandlerInterceptor {
+
+  public static final String EVENT_NAME = "drshub:api";
+
+  @Override
+  public void afterCompletion(
+      HttpServletRequest request,
+      @NotNull HttpServletResponse response,
+      @NotNull Object handler,
+      Exception ex) {
+    var path = request.getRequestURI();
+    var responseStatus = HttpStatus.valueOf(response.getStatus());
+    // Note: invalid responses will not be logged since there are no sessions to associate users
+    // with
+    if (handler instanceof HandlerMethod handlerMethod
+        && handlerMethod.getMethod().getAnnotation(TrackCall.class) != null
+        && responseStatus.is2xxSuccessful()) {
+      var bearerToken = bearerTokenFactory.from(request);
+      var properties =
+          new HashMap<String, Object>(
+              Map.of("statusCode", responseStatus.value(), "requestUrl", path));
+
+      properties.putAll(readBody(request));
+
+      addToPropertiesIfPresentInHeader(request, properties, "x-user-project", "userProject");
+      addToPropertiesIfPresentInHeader(request, properties, "X-Forwarded-For", "ip");
+      addToPropertiesIfPresentInHeader(
+          request, properties, "drshub-force-access-url", "forceAccessUrl");
```

Comment:
I am missing the intention behind pulling these specific header values into properties, but I think comment(s) would help (or breaking the code out further).

Reply:
Oh that's because those are the only known headers (that I could find) and I didn't want to just pass along all headers since that could include some sensitive info (like auth tokens, etc.)

Reply:
Got it!  I think this would help grok-ability long-term to have in a code comment or something like that.